### PR TITLE
[SPARK-42873][SQL] Define Spark SQL types as keywords

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to define Spark SQL types as keywords.

### Why are the changes needed?
The non-keywords types cause some inconveniences while analysing/transforming the lexer tree. For example, while forming the stable column aliases, see https://github.com/apache/spark/pull/40126.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running the modified test suites:
```
$ build/sbt "test:testOnly *.ResolveAliasesSuite"
$ build/sbt "test:testOnly *.ParserUtilsSuite"
```

### CODE REVIEW

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 1]
```diff
@@ -993,14 +993,34 @@ colPosition
...
+    | TIMESTAMP | TIMESTAMP_NTZ | TIMESTAMP_LTZ
+    | STRING
+    | CHARACTER | CHAR
+    | VARCHAR
+    | BINARY
+    | DECIMAL | DEC | NUMERIC
+    | VOID
+    | INTERVAL
+    | ARRAY | STRUCT | MAP
+    ;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1014]: shall we still allow identifier? then we can keep the previous nice error message for supported type string.

Comment (MaxGekk) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1014]: `identifier` doesn't catch any bad input as it is defined:
```
IDENTIFIER
    : (LETTER | DIGIT | '_')+
    ;
```

Even on Spark 3.3, sometimes we get `Syntax error`:
```sql
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.2
      /_/

scala> sql("CREATE TABLE tbl(a #abc)")
org.apache.spark.sql.catalyst.parser.ParseException:
Syntax error at or near '#': extra input '#'(line 1, pos 19)

== SQL ==
CREATE TABLE tbl(a #abc)
-------------------^^^
```
or sometimes:
```sql
scala> sql("CREATE TABLE tbl(a 1abc)")
org.apache.spark.sql.catalyst.parser.ParseException:
DataType 1abc is not supported.(line 1, pos 19)
```

Comment (cloud-fan) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1014]: The previous code uses `identifier`, if we keep it, we can at least  be on-par with the previous error message.

Comment (MaxGekk) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1015]: ok. done.

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Highly likely, the GA `continuous-integration/appveyor/pr` is not related to my changes. I am going to merge this PR.

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @cloud-fan for review.

**Pull request closed without merging**